### ISSUE TITLE
[R] base_margin not working in a Multiclass Cross-validation

### ISSUE THREAD

**Issue by brebbles** [dmlc/xgboost#5288]
Firstly, I see this topic has been previously raised/closed here [#2006](https://github.com/dmlc/xgboost/issues/2006), but I am still having the issue when running a multiclass classifier.  For reference I am using xgBoost version 1.0.0.1 via R:

I cannot get the base_margin passed through during cross-validation for a multiclass model.  I can get it working on xgb.train using the exact same code (save for the **nfold** argument), but the base_margin is not pass to the cross-validation run.

Further, the base_margin does pass through to the Cross Validation when when I run a standard binary classifier using the agaricus data (I can provide my code for this if needed).  It is just the multiclass scenario that doesn't work.

I have provided some code that should hopefully be replicated for the multi-class case.  

    #load xgboost
    library(xgboost)
     
    #define number of classes, features and label
    num_class <- 3
    data = as.matrix(iris[, -5])
    label =  as.numeric(iris$Species) - 1

    #create xgbMatrix
    xgbMat <- xgb.DMatrix(data = data, label = label)
    watchlist = list(train = xgbMat)

    #run initial boost
    bst <- xgb.train(data = xgbMat,
                     max_depth = 4, eta = 0.4, nrounds = 50,
                     objective = "multi:softprob", eval_metric = "mlogloss",
                     watchlist = watchlist,
                     num_class = num_class, print_every_n = 10)

That achieves the following multiclass log-loss metrics:

    [1]  train-mlogloss:0.636670 
    [11] train-mlogloss:0.039032 
    [21] train-mlogloss:0.022351 
    [31] train-mlogloss:0.018615 
    [41] train-mlogloss:0.016476 
    [50] train-mlogloss:0.015429

So the best log-loss is 0.015429

Now I predict from this boost and use the margin output as the base margin in a new data matrix:

    #output margin from initial boost
    pred <- predict(bst, xgbMat, outputmargin = TRUE)

    #create xgbMatrix with base margin from initial boost
    xgbMatPred <- xgb.DMatrix(data = data,label = label, base_margin = as.matrix(pred))
    watchlist = list(train = xgbMatPred)

Training using the new matrix with the base margin, you can see it picks up where the previous boost finished in terms of log-loss:

    #boost from initial prediction
    bst <- xgb.train(data = xgbMatPred,
                     max_depth = 4, eta = 0.4, nrounds = 50,
                     objective = "multi:softprob", eval_metric = "mlogloss",
                     watchlist = watchlist,
                     num_class = num_class, print_every_n = 10)

    [1]  train-mlogloss:0.015345 
    [11] train-mlogloss:0.014510 
    [21] train-mlogloss:0.013976 
    [31] train-mlogloss:0.013575 
    [41] train-mlogloss:0.013296 
    [50] train-mlogloss:0.013115

However, when I use the same new matrix and parameters in a cross-validation, it seems to ignore the base margin and just starts again from scratch:

    #attempt CV from initial prediction
    set.seed(1)
    bstCV <- xgb.cv(data = xgbMatPred,
                  nfold = 5,              
                  max_depth = 4, eta = 0.4, nrounds = 50,
                  objective = "multi:softprob", eval_metric = "mlogloss",
                  # watchlist = watchlist, 
                  showsd = FALSE,
                  num_class = num_class, print_every_n = 10)

    [1]  train-mlogloss:0.642223|test-mlogloss:0.660639
    [11] train-mlogloss:0.043816|test-mlogloss:0.158496 
    [21] train-mlogloss:0.024552|test-mlogloss:0.182579 
    [31] train-mlogloss:0.020419|test-mlogloss:0.199851 
    [41] train-mlogloss:0.018195|test-mlogloss:0.211206 
    [50] train-mlogloss:0.017033|test-mlogloss:0.217913

The log-loss in round 1 doesn't carry on from the initial boost - rather it resets to no base margin at all.

I've been trying numerous alternatives to this (removing the watchlist parameter, changing between multi:softprob and multi:softmax, different eval_metrics, uninstalling/hard-deleting versions of xgboost).  Nothing seems to work.

I'm not sure why this issue would be happening as [#2006](https://github.com/dmlc/xgboost/issues/2006) has fixed the issue, at least for the binary scenario.  Why it would not work for the multiclass I am not sure of.

Kind regards,

**Comment by trivialfis** [comment]
The bug is in DMatrix slice.

**Comment by trivialfis** [comment]
@hcho3 Does it make sense to have `base_margin` for survival training?

**Comment by hcho3** [comment]
@trivialfis Do you mean `survival:aft`? Yes, since the tree ensemble will predict log survival time.

**Comment by trivialfis** [comment]
@hcho3 Em ..  I believe the `cv` method for survival is also broken as the `slice` doesn't handle `upper_bound` nor `lower_bound`.
